File: examples/lab_radon_example.cpp
/*
Copyright (c) 2024-2026 Sebastian Sassi

Permission is hereby granted, free of charge, to any person obtaining a copy of 
this software and associated documentation files (the "Software"), to deal in 
the Software without restriction, including without limitation the rights to 
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies 
of the Software, and to permit persons to whom the Software is furnished to do 
so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all 
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE 
SOFTWARE.
*/
#include <format>
#include <fstream>
#include <print>
#include <vector>

#include "zest/zernike_glq_transformer.hpp"

#include "celestial.hpp"
#include "linalg.hpp"
#include "matrix.hpp"
#include "time.hpp"
#include "zebra_angle_integrator.hpp"

namespace
{

constexpr double reduced_mass(double m1, double m2)
{
    return m1*m2/(m1 + m2);
}

std::vector<double> calculate_vmin(
    double nucleus_mass, double dm_mass, double vesc, [[maybe_unused]] double vdisp, double emax)
{
    const double emin = 0.0;
    const double red_mass = reduced_mass(dm_mass, nucleus_mass);

    std::size_t count = 50;
    std::vector<double> vmin(count);
    for (std::size_t i = 0; i < count; ++i)
    {
        const double energy = emin + (emax - emin)*(double(i)/double(count - 1UL));
        vmin[i] = std::sqrt(nucleus_mass*energy/(2.0*red_mass*red_mass));
        vmin[i] /= vesc;
    }

    return vmin;
}

// Left distribution unnormalized in this toy example
constexpr double distribution_norm([[maybe_unused]] double vdisp, [[maybe_unused]] double vesc)
{
    return 1.0;
}

[[maybe_unused]]
void print_to_file(const char* fname, zest::DynamicMDSpan<const double, 2> array)
{
    std::ofstream fs(fname);
    for (std::size_t i = 0; i < array.extents()[0]; ++i)
    {
        for (std::size_t j = 0; j < array.extents()[1]; ++j)
            fs << std::format("{:.16e} ", array[i, j]);
        fs << '\n';
    }
    fs.close();
}

void print_to_stdout(zest::DynamicMDSpan<const double, 2> array)
{
    for (std::size_t i = 0; i < array.extent(0); ++i)
    {
        for (std::size_t j = 0; j < array.extent(1); ++j)
            std::print("{:.16e} ", array[i, j]);
         std::println("");
    }
}

zest::DynamicMDArray<double, 2> radon_transform(
    std::size_t dist_order, double vesc, double vdisp, double dm_mass,
    double nucleus_mass, std::span<double> times, double emax)
{
    const double dist_norm = distribution_norm(vdisp, vesc);
    auto velocity_distribution = [&](const std::array<double, 3>& v)
    {
        const double inv_vdisp = 1.0/vdisp;
        constexpr zdm::la::Matrix<double, 3, 3> sigma = {
            3.0, 1.4, 0.5,
            1.4, 0.3, 2.1,
            0.5, 2.1, 1.7
        };
        return dist_norm*std::exp(-0.5*zdm::la::quadratic_form(sigma, v)*(inv_vdisp*inv_vdisp));
    };

    const double lon = 0.0;
    const double lat = 1.5;
    const double vcirc = vdisp;

    zdm::celestial::GCStoHCS gcs_to_hcs{lon, lat, vcirc};

    std::vector<zdm::la::Vector<double, 3>> vlab{times.size()};
    for (std::size_t i = 0; i < times.size(); ++i)
        vlab[i] = zdm::celestial::transform_velocity(gcs_to_hcs, times[i]);

    const std::vector<double> vmin
        = calculate_vmin(nucleus_mass, dm_mass, vesc, vdisp, emax);

    // zebradm
    // zest
    const zest::zt::ZernikeExpansion dist_expansion
        = zest::zt::ZernikeTransformerNormalGeo{}.forward_transform(
            velocity_distribution, vesc, dist_order);

    zest::DynamicMDArray<double, 2> out{vlab.size(), vmin.size()};

    zdm::zebra::AngleIntegrator<zdm::DistType::aniso, zdm::RespType::iso>
    radon_integrator(dist_order);

    radon_integrator.integrate((typename decltype(dist_expansion)::const_view)(dist_expansion), vlab, vmin, out);

    for (auto& element : out.flatten())
        element *= vesc*vesc;

    return out;
}

} // namespace

int main([[maybe_unused]] int argc, char** argv)
{
    const double vmax = atof(argv[1]);
    const double vdisp = atof(argv[2]);
    const std::size_t dist_order = std::size_t(atoi(argv[3]));
    const double dm_mass = atof(argv[4]);
    const double nucleus_mass = atof(argv[5]);
    const std::string_view start_date{argv[6]};
    const std::string_view end_date{argv[7]};
    const std::size_t ntime = std::size_t(atoi(argv[8]));
    const double emax = atof(argv[9]);

    std::vector<double> times = zdm::time::ut1_interval<zdm::time::j2000_utc>(start_date, end_date, ntime, "%Y-%m-%d").value();

    const auto out = radon_transform(dist_order, vmax, vdisp, dm_mass, nucleus_mass, times, emax);

    //const char* fname = "lab_radon_example.dat";
    //print_to_file(fname, out);
    print_to_stdout((typename decltype(out)::const_view)(out));
}

File: examples/lab_radon_example.py
# Copyright (c) 2024 Sebastian Sassi
#
# Permission is hereby granted, free of charge, to any person obtaining a copy 
# of this software and associated documentation files (the "Software"), to deal 
# in the Software without restriction, including without limitation the rights 
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell 
# copies of the Software, and to permit persons to whom the Software is 
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in 
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR 
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE 
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER 
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE 
# SOFTWARE.
#
# This example showcases the performance of the Zernike-based Radon transform
# in an interactive way. For every change of the parameters, it calls the C++
# program `lab_radon_example` (found in `lab_radon_example.cpp`, which needs to
# be compiled first for this program to work), which computes a simplified toy
# version of the dark matter direct detection event rate.
import argparse
import subprocess
import io
import functools

import numpy as np
import matplotlib.pyplot as plt
import matplotlib.widgets as widgets
import scipy.integrate as integrate

amu_to_gev = 0.9315

class RadonPlotter:
    def __init__(self):
        self.fig = plt.figure()
        self.ax1 = self.fig.add_subplot(2, 1, 1)
        self.ax2 = self.fig.add_subplot(2, 1, 2)

        self.ax1.callbacks.connect("xlim_changed", self.update_ax)
        self.ax2.callbacks.connect("xlim_changed", self.update_ax)

        self.ax_vmax = self.fig.add_axes([0.20, 0.15, 0.60, 0.03])
        self.ax_vdisp = self.fig.add_axes([0.20, 0.1, 0.60, 0.03])
        self.ax_dist_order = self.fig.add_axes([0.20, 0.05, 0.60, 0.03])
        self.ax_dm_mass = self.fig.add_axes([0.20, 0.0, 0.60, 0.03])

        self.ax_xe = self.fig.add_axes([0.95, 0.0, 0.05, 0.03])
        self.ax_ar = self.fig.add_axes([0.95, 0.05, 0.05, 0.03])
        self.ax_ge = self.fig.add_axes([0.95, 0.1, 0.05, 0.03])
        self.ax_si = self.fig.add_axes([0.95, 0.15, 0.05, 0.03])
        self.ax_c = self.fig.add_axes([0.95, 0.2, 0.05, 0.03])
        self.ax_f = self.fig.add_axes([0.95, 0.25, 0.05, 0.03])
        self.ax_na = self.fig.add_axes([0.95, 0.3, 0.05, 0.03])
        self.ax_i = self.fig.add_axes([0.95, 0.35, 0.05, 0.03])

        #self.fig.add_axes()

        self.nucleus_mass = {
            "Xe": amu_to_gev*131.293,
            "Ar": amu_to_gev*39.963,
            "Ge": amu_to_gev*73.630,
            "Si": amu_to_gev*28.085,
            "C": amu_to_gev*12.011,
            "F": amu_to_gev*18.998,
            "Na": amu_to_gev*22.990,
            "I": amu_to_gev*126.90
        }

        self.vmax = 537
        self.vdisp = 233
        self.dist_order = 30
        self.dm_mass = 1.0
        self.element = "Xe"
        self.tmin = 8200
        self.tmax = 8565
        self.emin = 0.0
        self.emax = 200
        self.energies = np.linspace(0.0, self.emax, 50)
        self.times = np.linspace(self.tmin, self.tmax, 24)

        self.slider_vmax = widgets.Slider(
            ax=self.ax_vmax, label=r"$v_{esc}$", valmin=500, valmax=600, valinit=self.vmax)
        self.slider_vdisp = widgets.Slider(
            ax=self.ax_vdisp, label=r"$v_0$", valmin=200, valmax=300, valinit=self.vdisp)
        self.slider_dist_order = widgets.Slider(
            ax=self.ax_dist_order, label=r"$L$", valmin=2, valmax=50, valinit=self.vdisp, valstep=1)
        self.slider_dm_mass = widgets.Slider(
            ax=self.ax_dm_mass, label=r"$m_{DM}$", valmin=0.4, valmax=2.0, valinit=self.dm_mass)
        
        self.slider_vmax.on_changed(self.update_vmax)
        self.slider_vdisp.on_changed(self.update_vdisp)
        self.slider_dist_order.on_changed(self.update_dist_order)
        self.slider_dm_mass.on_changed(self.update_dm_mass)
        
        self.button_xe = widgets.Button(self.ax_xe, "Xe")
        self.button_ar = widgets.Button(self.ax_ar, "Ar")
        self.button_ge = widgets.Button(self.ax_ge, "Ge")
        self.button_si = widgets.Button(self.ax_si, "Si")
        self.button_c = widgets.Button(self.ax_c, "C")
        self.button_f = widgets.Button(self.ax_f, "F")
        self.button_na = widgets.Button(self.ax_na, "Na")
        self.button_i = widgets.Button(self.ax_i, "I")

        self.button_xe.on_clicked(functools.partial(self.update_element, "Xe"))
        self.button_ar.on_clicked(functools.partial(self.update_element, "Ar"))
        self.button_ge.on_clicked(functools.partial(self.update_element, "Ge"))
        self.button_si.on_clicked(functools.partial(self.update_element, "Si"))
        self.button_c.on_clicked(functools.partial(self.update_element, "C"))
        self.button_f.on_clicked(functools.partial(self.update_element, "F"))
        self.button_na.on_clicked(functools.partial(self.update_element, "Na"))
        self.button_i.on_clicked(functools.partial(self.update_element, "I"))

        self.rate_energy = self.compute_radon()
        self.rate = integrate.simpson(self.rate_energy, self.energies)

        self.pcmesh = self.ax1.pcolormesh(self.rate_energy, vmin=0.0)
        self.line, = self.ax2.plot(self.times, self.rate)
    
    def update_vmax(self, val):
        self.vmax = val
        self.update()

    def update_vdisp(self, val):
        self.vdisp = val
        self.update()

    def update_dist_order(self, val):
        self.dist_order = val
        self.update()

    def update_dm_mass(self, val):
        self.dm_mass = val
        self.update()

    def update_element(self, element, event):
        self.element = element
        self.update()
    
    def update_tmin(self, val):
        self.tmin = val
        self.times = np.arange(self.tmin, self.tmax, 15)
        self.update()
    
    def update_tmax(self, val):
        self.tmax = val
        self.times = np.arange(self.tmin, self.tmax, 15)
        self.update()

    def update_ax(self, ax):
        self.tmin, self.tmax = ax.get_xlim()
        self.emin, self.emax = ax.get_ylim()
        self.emin = 0.0
        self.emax = max(0.0, self.emax)
        self.update()
    
    def update(self):
        if (self.emin < self.emax):
            self.rate_energy = self.compute_radon()
            self.rate = integrate.simpson(self.rate_energy, self.energies)

            self.pcmesh.set_array(self.rate_energy.ravel())
            self.line.set_data(self.times, self.rate)
            self.fig.canvas.draw()
    
    def compute_radon(self):
        p = subprocess.run(f"./lab_radon_example {self.vmax} {self.vdisp} {int(self.dist_order)} {self.dm_mass} {self.nucleus_mass[self.element]} {self.tmin} {self.tmax} {self.emax}", capture_output=True, text=True, shell=True)
        return np.loadtxt(io.StringIO(p.stdout))


if __name__ == "__main__":
    plotter = RadonPlotter()
    plt.show()
